src/remote_tables/pushdown_visitor.rs

```diff
@@ -0,0 +1,204 @@
+use datafusion::common::DataFusionError;
+use datafusion::error::Result;
+use datafusion::scalar::ScalarValue;
+use datafusion_expr::expr_visitor::{ExprVisitable, ExpressionVisitor, Recursion};
+use datafusion_expr::{BinaryExpr, Expr, Operator};
+
+pub struct FilterPushdown<T: FilterPushdownVisitor> {
+    pub source: T,
+    pub pushdown_supported: bool,
+    // LIFO stack for keeping the intermediate SQL expression results to be used in interpolation
+    // of the parent nodes. After a successful visit, it should contain exactly one element, which
+    // represents the complete SQL statement corresponding to the given expression.
+    pub sql_exprs: Vec<String>,
+}
+
+pub struct PostgresFilterPushdown {}
+pub struct SQLiteFilterPushdown {}
+pub struct MySQLFilterPushdown {}
+
+impl FilterPushdownVisitor for PostgresFilterPushdown {}
+
+impl FilterPushdownVisitor for SQLiteFilterPushdown {
+    fn op_to_sql(&self, op: Operator) -> Option<String> {
+        match op {
+            Operator::RegexMatch
+            | Operator::RegexIMatch
+            | Operator::RegexNotMatch
+            | Operator::RegexNotIMatch
+            | Operator::BitwiseXor => None,
+            _ => Some(op.to_string()),
+        }
+    }
+}
+
+impl FilterPushdownVisitor for MySQLFilterPushdown {
+    fn op_to_sql(&self, op: Operator) -> Option<String> {
+        match op {
+            // TODO: see if there's a way to convert the non-case sensitive match
+            Operator::RegexIMatch | Operator::RegexNotIMatch => None,
+            Operator::RegexMatch => Some("RLIKE".to_string()),
+            Operator::RegexNotMatch => Some("NOT RLIKE".to_string()),
+            Operator::BitwiseXor => Some("^".to_string()),
+            _ => Some(op.to_string()),
+        }
+    }
+}
+
+pub trait FilterPushdownVisitor {
+    fn scalar_value_to_sql(&self, value: &ScalarValue) -> Option<String> {
+        match value {
+            ScalarValue::Utf8(Some(val)) => Some(format!("'{}'", val)),
+            _ => Some(format!("{}", value)),
```

Comment:
I think some non-UTF scalar values still need to be in single quotes, e.g. datetimes in PostgreSQL. More here: https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-CONSTANTS

Perhaps we could make this safer by always emitting a cast (e.g. `'2022-01-01 12:34:56'::timestamp`) but then we'd need to know the database-specific type information.

Reply:
Follow-up: maybe there's some way we could use an ORM/SQL query builder in `sqlx` for a particular dialect to get around some database-level syntax differences.

Reply:
Oh yeah, no need to overcomplicate this PR by adding an ORM.